### PR TITLE
[APIDiff] Add baseline management flags

### PR DESCRIPTION
**Pull request by owenv**
Add flags that can be used to manage API baselines

### Motivation:

In some environments, it can be useful to check API baseline files into source control, those of tagged versions for example.

### Modifications:

- `--baseline-dir` can be used to choose a directory to emit API baselines in
- `--regenerate-baseline` wil cause the tool to ignore preexisting baseline files

### Result:

It's now possible to choose a directory to emit baseline file to and then check them in

### CODE REVIEW

**Comment by owenv** [comment]
@swift-ci please smoke test

**Comment by owenv** [comment]
@swift-ci please smoke test

**Comment by owenv** [comment]
@swift-ci please smoke test macOS

File: Sources/Commands/APIDigester.swift [hunk 1]
```diff
@@ -64,18 +64,22 @@ struct APIDigesterBaselineDumper {
     }
 
     /// Emit the API baseline files and return the path to their directory.
-    func emitAPIBaseline(for modulesToDiff: Set<String>) throws -> AbsolutePath {
+    func emitAPIBaseline(for modulesToDiff: Set<String>,
+                         baselineDir: AbsolutePath?,
+                         forceRegenerateBaseline: Bool) throws -> AbsolutePath {
```
Comment (elsh) [inline Sources/Commands/APIDigester.swift:69]: To be more succinct, 
```
at baselineDir: AbsolutePath?,
force: Bool)
```

**Review by elsh — COMMENTED** [review]
I agree with giving an option to override the baseline dir, but _not_ with checking in the result files; those should be kept separately from the source control.

File: Sources/Commands/APIDigester.swift (hunk 1 above)
Comment (owenv) [reply Sources/Commands/APIDigester.swift]: 👍 updated

**Comment by owenv** [comment]
> I agree with giving an option to override the baseline dir, but not with checking in the result files; those should be kept separately from the source control.

yeah, the use cases I originally had in mind for checking in baselines haven't worked out in practice (especially because they embed compiler flags and platform-specific info), but I agree this is otherwise a generally useful option to provide

**Comment by owenv** [comment]
@swift-ci please smoke test

File: Sources/Commands/APIDigester.swift [hunk 2]
```diff
@@ -64,18 +64,22 @@ struct APIDigesterBaselineDumper {
...
         let baselinePath: (String)->AbsolutePath = { module in
             baselineDir.appending(component: module + ".json")
         }
 
-        for module in modulesToDiff {
-            if localFileSystem.exists(baselinePath(module)) {
-                // If this baseline already exists, we don't need to regenerate it.
-                modulesToDiff.remove(module)
+        if !force {
+            for module in modulesToDiff {
```
Comment (elsh) [inline Sources/Commands/APIDigester.swift:78]: You can also use ```.filter``` instead; ```remove``` can incur a perf cost.

File: Sources/Commands/APIDigester.swift [hunk 3]
```diff
@@ -64,18 +64,22 @@ struct APIDigesterBaselineDumper {
     }
 
     /// Emit the API baseline files and return the path to their directory.
-    func emitAPIBaseline(for modulesToDiff: Set<String>) throws -> AbsolutePath {
+    func emitAPIBaseline(for modulesToDiff: Set<String>,
+                         baselineDir: AbsolutePath?,
```
Comment (elsh) [inline Sources/Commands/APIDigester.swift:68]:
```suggestion
                         at baselineDir: AbsolutePath?,
```

**Review by elsh — APPROVED** [review]

File: Sources/Commands/APIDigester.swift [hunk 4]
```diff
@@ -64,18 +64,22 @@ struct APIDigesterBaselineDumper {
...
 
-        for module in modulesToDiff {
-            if localFileSystem.exists(baselinePath(module)) {
-                // If this baseline already exists, we don't need to regenerate it.
-                modulesToDiff.remove(module)
+        if !force {
+            for module in modulesToDiff {
+                if localFileSystem.exists(baselinePath(module)) {
+                    // If this baseline already exists, we don't need to regenerate it.
+                    modulesToDiff.remove(module)
```
Comment (elsh) [inline Sources/Commands/APIDigester.swift:81]: You can also use .filter instead (of this for loop); remove can incur a perf cost.

**Comment by owenv** [comment]
@swift-ci smoke test

**Comment by owenv** [comment]
@swift-ci please smoke test macOS

**Comment by tomerd** [comment]
@owenv @elsh is this ready to be merged?

**Pull request merged**